test/testdrive/materialization-lag.td

```diff
@@ -0,0 +1,185 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Test the contents of `mz_materialization_lag`.
+
+> CREATE CLUSTER source SIZE '1'
+> CREATE CLUSTER compute SIZE '1'
+> CREATE CLUSTER sink SIZE '1'
+
+# Set up a bunch of sources and materializations that depend on each other.
+# Put them on different clusters so we can spin some down to create lag.
+
+> CREATE SOURCE src
+  IN CLUSTER source
+  FROM LOAD GENERATOR counter
+
+> CREATE TABLE tbl (a int)
+
+> CREATE VIEW src_plus_tbl
+  AS SELECT counter + a AS a FROM src, tbl
+
+> CREATE INDEX idx
+  IN CLUSTER compute
+  ON src_plus_tbl (a)
+
+> CREATE MATERIALIZED VIEW mv
+  IN CLUSTER compute
+  AS SELECT * FROM src_plus_tbl
+
+> CREATE CONNECTION kafka_conn
+  TO KAFKA (BROKER '${testdrive.kafka-addr}')
+> CREATE CONNECTION csr_conn
+  TO CONFLUENT SCHEMA REGISTRY (URL '${testdrive.schema-registry-url}')
+> CREATE SINK snk
+  IN CLUSTER sink
+  FROM mv
+  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink1-${testdrive.seed}')
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+  ENVELOPE DEBEZIUM
+
+# Make sure things are hydrated.
+
+> INSERT INTO tbl VALUES (1)
+> SELECT count(*) > 0 FROM mv
+true
+
+# When all clusters are running, there should be no visible lag.
+
+> SELECT name
+  FROM mz_internal.mz_materialization_lag
+  JOIN mz_objects ON (id = object_id)
+  WHERE
+    local_lag > INTERVAL '1s' OR
+    global_lag > INTERVAL '1s'
+
+# When the source is down, there should be no visible lag either, as lag is
+# relative to the source frontiers.
+
+> ALTER CLUSTER source SET (REPLICATION FACTOR 0)
+
+> SELECT name
+  FROM mz_internal.mz_materialization_lag
+  JOIN mz_objects ON (id = object_id)
+  WHERE
+    local_lag > INTERVAL '1s' OR
+    global_lag > INTERVAL '1s'
+
+> ALTER CLUSTER source SET (REPLICATION FACTOR 1)
+
+# Bring down the compute cluster and observe resulting lag.
+
+> ALTER CLUSTER compute SET (REPLICATION FACTOR 0)
+
```

Comment:
Do we need a call to `mz_sleep` here to avoid flakiness?

Reply:
Retries are considered table stakes for correctness in Testdrive IMO, I'm pretty sure we already have a huge amount of tests that will fail with retries turned off.

Reply:
In that case, feel free to send it.

Reply:
Yeah, most of the tests that query introspection relations include a blurb at the beginning of the .td file saying that this relies on testdrive's retries feature. I'll add that here too.